Add ConvNeXt

sayakpaul opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons --sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Closes https://github.com/keras-team/keras-cv/issues/523


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? https://github.com/keras-team/keras-cv/issues/523
- [x] Did you write any new necessary tests?

Here's my playground Colab: https://colab.research.google.com/gist/sayakpaul/a0edc7f31f1402274126b37a543ea272/scratchpad.ipynb


## Who can review?

@LukeWood @ianstenbit @tanzhenyu

<!-- 
Feel free to tag @LukeWood and @tanzhenyu in your reviews.
-->

<!--
This PR template is copied and modified from here:
https://github.com/huggingface/transformers/blob/main/.github/PULL_REQUEST_TEMPLATE.md
-->

sayakpaul commented:
@LukeWood @ianstenbit @tanzhenyu

ready for review.

bhack commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+    # Stochastic depth schedule.
+    # This is referred from the original ConvNeXt codebase:
+    # https://github.com/facebookresearch/ConvNeXt/blob/main/models/convnext.py#L86
+    depth_drop_rates = [float(x) for x in np.linspace(0.0, drop_path_rate, sum(depths))]
```
Can we remove the numpy dependency in non-testing code with native TF ops?

https://github.com/tensorflow/tensorflow/blob/v2.9.1/tensorflow/python/ops/numpy_ops/np_math_ops.py#L1049-L1090

sayakpaul replied:
I remember using the native TF ops failed in an internal test when I contributed the same model to `keras.applications`. 

Cc: @LukeWood

ianstenbit replied:
Could you please demonstrate that our tests fail if using a native TF op here?
I agree that we should prefer a native TF op.

sayakpaul replied:
Addressed it.

sayakpaul commented:
The tests run successfully on my end.

This is what I used:

```
pytest keras_cv/models/models_test.py
```

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+          be a 2D tensor.
+        - `max` means that global max pooling will be applied.
+      classes: optional number of classes to classify images into, only to be
+        specified if `include_top` is True, and if no `weights` argument is
```
Since we're not inferring `classes` based on `weights`, I think we should drop this last clause.

I'm making this change for other models in #689

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+    projection_dims,
+    drop_path_rate=0.0,
+    layer_scale_init_value=1e-6,
+    include_top=True,
```
Let's make this a required parameter, even for the internal method, for consistency with other models and clarity that this is something the API expects from the user.

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+    input_shape=(None, None, 3),
+    input_tensor=None,
+    pooling=None,
+    classes=1000,
```
Let's default `classes=None`

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+        ValueError: if `classifier_activation` is not `softmax`, or `None`
+          when using a pretrained top layer.
+        ValueError: if `include_top` is True but `num_classes` is not 1000
+          when using ImageNet.
```
Instead of specifying 1000/imagenet, let's just say if `num_classes` is not specified.

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+    return apply
+
+
+def Head(num_classes=1000, activation="softmax", name=None):
```
Let's not include a default here -- this should be a required param, as we don't want to implicitly expect ImageNet

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,494 @@
+    include_rescaling,
+    drop_path_rate,
+    layer_scale_init_value,
+    include_top,
```
For API consistency, please make `include_rescaling` and `include_top` the first two required/positional parameters to each public ConvNeXt method

ianstenbit reviewed (changes requested):
Thank you! Excited to have ConvNeXt available in KerasCV.

Mostly minor comments for you, sorry for the delay.

sayakpaul commented:
@ianstenbit addressed your comments.

ianstenbit commented:
keras-cv-accelerator-testing is failing because #692 is not yet merged -- you can ignore that for now.

ianstenbit commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,493 @@
+    )
+
+
+def ConvNeXtXLarge(
```
Please add Docstrings for these exported methods using the convention from the DenseNet impl:
https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/densenet.py

sayakpaul replied:
Such an awesome piece of work that is!

ianstenbit reviewed (approved):
Thank you! Looks good to me after my one comment.

sayakpaul commented:
> Thank you! Looks good to me after my one comment.

Such an awesome piece of work that is! @ianstenbit

sayakpaul commented:
@ianstenbit addressed your comment.

Should `gs://keras-cv/models` be publicly viewable? I could see `gs://keras-cv/models` with `gsutil ls -lh gs://keras-cv/models`.

sayakpaul commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,539 @@
+            depth won't be used.
+        layer_scale_init_value: layer scale coefficient, if 0.0, layer scaling
+            won't be used.
+        weights:
```
Not sure what do I put here. `weights.py` doesn't have any weights for ConvNeXt.

ianstenbit replied:
For now, this will only support loading weights from a directory (no pre-trained weights).

If you'd like to train some weights, feel free to do so using basic_training.py or by contributing your own training script.

sayakpaul commented on keras_cv/models/convnext.py:
```diff
@@ -0,0 +1,539 @@
+        depth won't be used.
+      layer_scale_init_value: Layer scale coefficient. If 0.0, layer scaling
+        won't be used.
+      weights:
```
Same as https://github.com/keras-team/keras-cv/pull/678/files#r947004467.

ianstenbit replied:
(Same -- let's use the string "one of `None` (random initialization), or a pretrained weight file path")

sayakpaul replied:
Can you suggest a text?

ianstenbit replied:
Yes, sorry I wasn't clear.

Let's use "one of None (random initialization), or a pretrained weight file path", which is the same as the docstrings from the other models which don't yet support pre-trained weights.

ianstenbit commented:
> @ianstenbit addressed your comment.
> 
> Should `gs://keras-cv/models` be publicly viewable? I could see `gs://keras-cv/models` with `gsutil ls -lh gs://keras-cv/models`.

Yes this is WAI -- this is where we are storing weights so this needs to be viewable to the public

LukeWood commented on keras_cv/models/models_test.py:
```diff
@@ -69,7 +99,12 @@ def test_application_base(self, app, _, args):
         config = model.get_config()
-        reconstructed_model = model.__class__.from_config(config)
+        if "ConvNeXt" in app.__name__:
+            custom_objects = {"LayerScale": convnext.LayerScale}
```
will registering this with @tf.keras.utils.register_keras_serializable make us not need this?

sayakpaul replied:
Could you provide an example?

LukeWood replied:
https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/aug_mix.py#L24

Basically loading our models should not require custom object registration - so lets give this a shot.

sayakpaul commented:
@LukeWood @ianstenbit should be all set now.

LukeWood commented:
/gcbrun

LukeWood commented:
Looks like an OOM error 🤔

sayakpaul commented:
> Looks like an OOM error 🤔

Is there anything I can do to prevent that?

LukeWood commented:
Fixed tests, lets try pushing again to run them again

sayakpaul commented:
@LukeWood, `keras-cv-accelerator-testing (keras-team-test)` was broken as soon as I pushed.

LukeWood commented:
/gcbrun

LukeWood commented:
@sayakpaul I rebased, lets see if it passes and then we can merge

LukeWood commented:
/gcbrun

LukeWood commented:
@ianstenbit lets run the imagenet script on `ConvNeXtLarge` next!  See what scores we can produce.

For perspective, ConvNeXtLarge scores up to 85~% on ImageNet-22k beating Swin-Transformers, the other best model for image classification.  If we can get to 82% I'd be really happy with where our training script stands.

sayakpaul commented:
Why the rebasing was needed? Could have been just an empty commit no?

LukeWood commented:
/gcbrun

LukeWood commented:
/gcbrun

LukeWood commented:
/gcbrun
